### ISSUE TITLE
running multiple tasks on different ports broken

### ISSUE THREAD

**Issue by zy4** [playframework/playframework#9298]
Running multiple tasks on different ports used to work as follows:

```scala
val runAll = inputKey[Unit]("Runs main server and micro-service")

runAll := {
  (run in `microservice` in Compile).partialInput(" 9001").evaluated
  (run in root in Compile).partialInput(" 9003").evaluated
}
```

Since 2.7.1 / 2.7.2 the second task crashes with:

```
[error] a.i.TcpListener - Bind failed for TCP channel on endpoint [/0.0.0.0:9001]
java.net.BindException: Address already in use
	at sun.nio.ch.Net.bind0(Native Method)
	at sun.nio.ch.Net.bind(Net.java:433)
	at sun.nio.ch.Net.bind(Net.java:425)
	at sun.nio.ch.ServerSocketChannelImpl.bind(ServerSocketChannelImpl.java:223)
	at sun.nio.ch.ServerSocketAdaptor.bind(ServerSocketAdaptor.java:74)
	at akka.io.TcpListener.liftedTree1$1(TcpListener.scala:59)
	at akka.io.TcpListener.<init>(TcpListener.scala:56)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
```

Might be related to: https://github.com/playframework/playframework/pull/9099 (not sure). Running only one task works as usual. It only happens in DEV mode.

**Comment by mkurz** [comment]
Are you sure you did shutdown all running apps so nothing is blocking port 9001 when running `runAll`?

**Comment by zy4** [comment]
yes

**Comment by mkurz** [comment]
When changing it randomly then it's the random port that is blocking? Or always 9001?

**Comment by zy4** [comment]
The first one.

**Comment by zy4** [comment]
I also tried to not pass the port as an input parameter to the input task but to configure it as 
`play.http.port = ... ` in an application.conf in the respective sub projects.
Then always port 9000 is used, also with the blocking behaviour.

**Comment by mkurz** [comment]
Please try `-D...` syntax:
```scala
val runAll = inputKey[Unit]("Runs main server and micro-service")

runAll := {
  (run in `microservice` in Compile).partialInput(" -Dplay.server.http.port=9001").evaluated
  (run in root in Compile).partialInput(" -Dplay.server.http.port=9003").evaluated
}
```

**Comment by zy4** [comment]
This seems to work. Should we close the issue? Since this is a breaking change (there are examples of the version without -D, even in books).

**Comment by mkurz** [comment]
What happens here is a result of fixing a couple of issues:

Until now `-Dplay.server.http.port` was [_not_ taken into account](https://github.com/playframework/playframework/blob/ad7ed0c2c0e107a183d47885592044926b38a833/framework/src/run-support/src/main/scala/play/runsupport/Reloader.scala#L104) when supplied via command line, _but_ it was taken into account when setting it via `PlayKeys.devSettings` in `build.sbt`. #9096 fixed that inconsistency. (Also it's not just an inconsistency, but was a real bug because you could set every other config via `-D...` except the port and binding address until that fix).

Plus, important, when passing the port like you did in your initial comment by just adding the port number to the `run` command, Play now also sets the `play.server.http.port` java property - this makes sure that a developer can rely on the fact that when reading the config `play.server.http.port` it always correctly supplies the port the application is actually running on - even in dev mode. So passing the port number to `run` is just a shortcut for setting `play.server.http.port`. So this was a necessary fix IMHO and was done in #9298.

[We also decided](https://github.com/playframework/playframework/pull/9096/files#r266082301) following order when determining the port:
https://github.com/playframework/playframework/blob/f28075466fa9734f8743b663f193207eb49f7171/dev-mode/run-support/src/main/scala/play/runsupport/Reloader.scala#L114-L119
As you can see the java property `play.server.http.port` always overrules everything else (this is consistent how Play play behaves accross the framework). So what now happens is that your first `run` command (`(run in microservice in Compile).partialInput(" 9001")`) will set the `play.server.http.port` java property in the JVM and the second command will then use this property to determine its port as well... because it's first in the ranking... and the command line argument is just third in the ranking (`otherArgs.headOption`).

I will close this because I think all the fixes are necessary bug fixes and we should not change it back. Plus I guess your setup is very specific and not many people will be effected by such a corner case. Also you know how it works now so all is good :wink:
